### ISSUE TITLE
Speculative decoding (`draft_model()`) + non-greedy sampling: crash and silent output corruption from a one-iteration-late rejection handler

### ISSUE THREAD

**Issue by Jermalk** [openvinotoolkit/openvino.genai#4190]
## Summary

I have stumbled on this issue during implementation of speculative decoding in my custom OpenVINO-based, Rust-written server. The issue was discovered, diagnosed, and patched on two of my four machines in the fleet, so it seems reproducible. Documents are processed and prepared with Claude Fable 5. I can deliver additional artifacts if needed and test on an Arc PRO B60 or Lunar Lake machine.

Running `ov::genai::draft_model()` speculative decoding (plain draft-model pairing, not EAGLE) with non-greedy sampling (`do_sample=True`, any `temperature > 0`) produces two symptoms of one defect in `Sampler::sample_from_sequence_group` (`src/cpp/src/sampling/sampler.cpp`):

1. Intermittent crash: `Check 'content_length <= prompt_ids.size() + m_generated_ids.size()' failed` in `Sequence::_make_hash()` (`src/cpp/src/sequence_group.cpp:32`).
2. With prefix caching disabled (which removes the code path containing that assertion), no crash — but generated text is silently corrupted compared to plain decoding at the same temperature: a word gets dropped and its predicted continuation kept in its place.

Both come from the same root cause: on a rejected draft candidate, the non-greedy path handles the rejection one loop iteration late, in a different branch, with an offset that is stale by then. One change — handle the rejection inline, in the iteration where it's detected, the way the greedy/prompt-lookup path already does — fixes both. I verified this by building from source; a crash-only partial fix does not stop the corruption (details below), so the two symptoms should not be patched independently.

## Environment

- `openvino_genai` 2026.2.0.0, tag `2026.2.0.0`, commit `adf73e80e66629730f976d44cad6c09cf978deca` (checked with `git describe --tags` that the commit is the tag, not a snapshot ahead of it)
- `openvino` core 2026.2.0.0
- Intel Arc GPU (Battlemage, `xe` driver, Linux), reproduced on two separate machines: Arc Pro B50 and Arc Pro B70
- Target/draft: dense Qwen3 chat models at two precisions — Qwen3-14B + Qwen3-0.6B, int4 on one machine, int8 on the other. Same tokenizer family. Plain `ContinuousBatchingPipeline` + `draft_model()`, not EAGLE/tree-search.

## Minimal reproduction (pure Python, `openvino_genai` only)

Any dense Qwen3 OpenVINO IR pair works. Convert with `optimum-cli`:

```sh
optimum-cli export openvino --model Qwen/Qwen3-14B --weight-format int4 ./qwen3-14b-int4-ov
optimum-cli export openvino --model Qwen/Qwen3-0.6B --weight-format int4 ./qwen3-0.6b-int4-ov
```

(Pre-quantized IRs from the `OpenVINO` org on Hugging Face also work.)

```python
import openvino as ov
import openvino_genai as ovg

TARGET_MODEL_DIR = "./qwen3-14b-int4-ov"
DRAFT_MODEL_DIR = "./qwen3-0.6b-int4-ov"
DEVICE = "GPU"  # reproduced on GPU; CPU not tried yet

sched = ovg.SchedulerConfig()
sched.max_num_seqs = 1
sched.cache_size = 4  # GB
sched.enable_prefix_caching = True  # set False to see the corruption without the crash

draft = ovg.draft_model(DRAFT_MODEL_DIR, DEVICE)
properties = {"draft_model": draft, "KV_CACHE_PRECISION": ov.Type.u8}
pipe = ovg.ContinuousBatchingPipeline(TARGET_MODEL_DIR, sched, DEVICE, properties=properties)
tok = pipe.get_tokenizer()

formatted_prompt = tok.apply_chat_template(
    [{"role": "user", "content": "Tell me a fact."}],
    add_generation_prompt=True,
    extra_context={"enable_thinking": False},
)

cfg = ovg.GenerationConfig()
cfg.max_new_tokens = 40
cfg.num_assistant_tokens = 1  # single candidate per step: every rejection is a
                              # last-candidate rejection, see root cause
cfg.do_sample = True
cfg.temperature = 0.8

for i in range(40):
    result = pipe.generate(formatted_prompt, cfg)  # intermittently raises the assertion
    print(i, result)
```

Trigger conditions: `draft_model()` attached, `do_sample=True`, any prompt. `num_assistant_tokens=1` makes the failure near-total; larger values fail intermittently and non-monotonically across identical-parameter calls. Greedy decoding (`temperature=0`) is unaffected at any `num_assistant_tokens`.

Observed rate with this exact loop, unpatched: roughly 2–7% of calls raise the assertion, varying with prompt and temperature; a batch of 40 reliably hits at least one. The exact rate matters less than that the assertion is reachable from ~15 lines of pure Python.

## Root cause

Non-greedy rejection handling in `sample_from_sequence_group` is split across two points, one loop iteration apart.

Point A — the iteration where the rejection is detected. A correctly-positioned replacement token has already been sampled from the target's logits for this position, then the code throws it away:

```cpp
if (!is_validation_passed && sampling_params.do_sample && !sampling_params.is_prompt_lookup()) {
    continue;   // discards the correct sample, moves to the next iteration
}
```

This branch never touches `assisting_pipeline_info.max_removed_tokens_per_request`.

Point B — the following iteration, which looks back at "the previous token was not accepted", resamples, and is the only place the rollback counter is set:

```cpp
if (!is_validation_passed) {
    float p_prime = get_p_prime(running_sequence, sampled_token, generated_seq_token_offset + 1);
    assisting_pipeline_info.max_removed_tokens_per_request =
        std::max(assisting_pipeline_info.max_removed_tokens_per_request, generated_seq_token_offset);
    // ... log-prob correction, doesn't change which token is chosen
}
```

Two consequences:

1. The crash. `generated_seq_token_offset` at Point B is one less than the rejected candidate's offset at Point A. With `C` candidates (offsets `C…1`), a rejection at offset `k >= 2` leaves the counter at `k-1 >= 1` and the post-`finish_iteration()` rollback gate (`if (assisting_pipeline_info.max_removed_tokens_per_request)`) fires. A rejection at offset `k = 1` — the last or only candidate — leaves the counter at 0, the gate never fires, and `m_num_processed_tokens` stays one token too high for the rest of the sequence. On the next scheduling step, `content_length` exceeds `prompt_ids.size() + m_generated_ids.size()` by exactly 1 when prefix caching's block hashing runs — the assertion. This is why `num_assistant_tokens=1` makes it near-total: every rejection is a last-candidate rejection.
2. The corruption, independent of the crash. The `continue` at Point A discards the correct sample and lets the next iteration draw a token from the logit row for the position after the rejected one, conditioned on the rejected draft token as context. That wrong-row token is what the client receives. Net effect: the rejected word is deleted and its predicted continuation kept, e.g. `"...covers than third of Earth surface..."` instead of `"...covers more than a third of Earth's surface..."`. Prefix caching only gates whether the assertion fires; the wrong-row sampling happens either way, which is why disabling it stops the crash but not the corruption.

The greedy/prompt-lookup path in `validate_candidate()` has no deferred branch — it removes the rejected tokens, sets the counter, and registers the replacement inline in the same iteration, which is why greedy was never affected. The fix applies the same pattern to the non-greedy path.

## The fix

One change, entirely at Point A — handle the rejection inline instead of deferring it:

- remove the rejected candidate from the sequence tail;
- set `max_removed_tokens_per_request` to this rejection's own offset (`generated_seq_token_offset`, no iteration lag) — this makes the gate always fire and fixes the crash;
- register the sample already drawn this iteration (the correctly-positioned one) instead of hitting `continue` — this fixes the corruption.

Point B's `p_prime` block becomes unreachable (nothing arrives there with `is_validation_passed == false` anymore) and is removed. Diff in the appendix.

A note on sampling fidelity: the removed `p_prime` block looks like an attempt at the acceptance-correction resample from canonical speculative sampling (drawing the replacement from the normalized residual distribution, which keeps the output distribution exactly equal to non-speculative sampling). As implemented it operates on the wrong iteration's logit row, so it can't be kept as-is. My patch replaces it with a direct draw from the target distribution at the rejected position — matching how the greedy path handles rejection, and indistinguishable from plain decoding in my content comparisons. If you'd rather have a distribution-exact residual resample at the correctly-positioned row, I'm happy to implement and verify that variant instead.

## Verification

All counts below are direct `ContinuousBatchingPipeline.generate()` calls through the Python API above.

Unpatched: assertion reproduced repeatedly across independent runs of the 40-call loop; with prefix caching off, no crash but corrupted output.

Gate-only partial fix (fixes only the Point-B counter lag, leaves the `continue`): crash gone, 40/40 clean — but a 10-prompt content comparison against plain decoding at matched temperature/seed still showed most non-trivial completions corrupted, e.g. `"Here's a fact:Oct have same falling of breath they have same falling of breath as do have water as same structure."` Same corruption with `enable_prefix_caching=False`. So a counter-only patch looks complete (no crashes) but ships corrupted output.

Full fix, Qwen3-14B-int8 + Qwen3-0.6B-int8 on the B70, 400-token generations, 8 varied prompts (short factual plus long-form essay/technical/narrative), every generation read manually:

- prefix caching on: 12/12 coherent, zero corruption;
- prefix caching off: 12/12 coherent, zero corruption.

An earlier, shorter run on the B50 with the int4 pairing gave the same result: 10/10 coherent, 40/40 crash-free in a separate crash-reproduction batch.

## Supplementary: corroboration through my own inference server

I maintain an OpenAI-compatible Rust inference server (RustedVINO) wrapping `ContinuousBatchingPipeline` directly, and used it as a second harness, including under concurrent load. Not needed to reproduce the bug; included because the results were clean and may be useful. Unpatched: 21/21 non-greedy requests at `num_assistant_tokens=1` failed, 3/3 greedy clean. Patched: 21/21 non-greedy clean, 3/3 greedy clean, and a concurrency ramp (2/4/8 simultaneous streams, prefix caching on, 3 rounds each) came back 6/6, 12/12, 24/24 clean, wall time flat across the ramp, no GPU/driver kernel errors (`journalctl -k` monitored throughout).

## Related prior art

PR #1951 ("Remove tokens after EOS for draft model for speculative decoding", merged March 2025, shipped in 2025.2) fixed a different case throwing the identical assertion text — draft tokens generated past an EOS marker, described for greedy decoding. My version is well past that fix; this looks like a distinct case in the same code area, not a regression of #1951.

## Note on OVMS

OpenVINO Model Server's EAGLE3 speculative-decoding docs list explicit limitations: concurrency capped at 1, prefix caching not supported, greedy-only sampling enforced. Two of those three match the variables this bug centers on. For the classic (non-EAGLE) pairing there is no documented restriction, but every example request in the docs, for either pairing type, uses `temperature=0`; I did not find a non-greedy example anywhere.

## Not tested yet

- Only Intel Arc (Battlemage) GPUs, two machines, same `xe` driver family — no CPU-only, NPU, or non-Intel backends.
- Only the Qwen3 family (14B target, 0.6B draft, int4 and int8) — no other target/draft family.
- Both defects are still present in current `master` as of this writing.

Full logs and scripts available on request; happy to run additional diagnostics.

## Appendix: the patch

Applies cleanly to `src/cpp/src/sampling/sampler.cpp` at the `2026.2.0.0` tag (`adf73e80e66629730f976d44cad6c09cf978deca`).

```diff
diff --git a/src/cpp/src/sampling/sampler.cpp b/src/cpp/src/sampling/sampler.cpp
--- a/src/cpp/src/sampling/sampler.cpp
+++ b/src/cpp/src/sampling/sampler.cpp
@@ -940,17 +940,10 @@ SequenceGroupSamplingInfo Sampler::sample_from_sequence_group(SequenceGroup::Ptr
                         sg_sampling_info.sampler_output.m_forked_sequences.insert({running_sequences[0]->get_id(), forked_seq_ids});
                     }
                     sampled_token = sampled_token_ids.front();
-                    // make `_speculative_sampling` in case of previous token was not accepted in speculative decoding
-                    if (!is_validation_passed) {
-                        float p_prime = get_p_prime(running_sequence, sampled_token, generated_seq_token_offset + 1);
-                        assisting_pipeline_info.max_removed_tokens_per_request = std::max(assisting_pipeline_info.max_removed_tokens_per_request, generated_seq_token_offset);
-                        // update prob only in case candidate prob > sampled token prob
-                        if (p_prime > 0.f) {
-                            auto prob = std::exp(sampled_token.m_log_prob);
-                            prob /= p_prime;
-                            sampled_token.m_log_prob = std::log(prob);
-                        }
-                    }
+                    // Rejections are now handled inline where detected (below), so
+                    // is_validation_passed is always true here. The old block resampled
+                    // from the next iteration's logit row (wrong position).
                 }
                 if (!is_validation_mode_enabled && m_d2t_mapping) { // compute token offset for draft model in speculative sampling
                     ov::Tensor d2t_tensor = m_d2t_mapping->get_tensor_view();
@@ -964,9 +957,13 @@ SequenceGroupSamplingInfo Sampler::sample_from_sequence_group(SequenceGroup::Ptr
                                                               sampled_token, is_extend_sequence, assisting_pipeline_info.max_removed_tokens_per_request,
                                                               sampling_params.do_sample, !sampling_params.is_prompt_lookup(), rng_engine);
 
-                    // doing resample in case of non accepted tokens in speculative sampling
+                    // Non-greedy rejection: use the sample already drawn at this position
+                    // (same pattern as the greedy path) instead of deferring to the next
+                    // iteration's wrong-position row.
                     if (!is_validation_passed && sampling_params.do_sample && !sampling_params.is_prompt_lookup()) {
-                        continue;
+                        running_sequence->remove_last_tokens(generated_seq_token_offset);
+                        assisting_pipeline_info.max_removed_tokens_per_request =
+                            std::max(assisting_pipeline_info.max_removed_tokens_per_request, generated_seq_token_offset);
+                        is_extend_sequence = true;
                     }
                     // update log prob just while validation process
                     if (!is_extend_sequence) {
```

Note: I shortened the in-diff comments for readability here; the PR will carry the same comments in the source, so hunk line counts in the final diff may differ slightly from the above.

---

AI assistance used: yes
How: investigation partner and drafting/editing of this report and the patch
Human validation performed: reproduced on my own hardware (Arc Pro B50 and B70), built openvino_genai from source at the tagged commit, ran all experiments and content comparisons myself, read every generation manually

**Comment by apaniukov** [comment]
Hi @Jermalk,
Thank you for the investigation and a detailed description!
